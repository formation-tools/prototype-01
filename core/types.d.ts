type FRecursiveObject = {
  content: FObjectType[]
}

type FLink = FRecursiveObject & {
  type: 'a'
  target: string
  linkType: string
}

type FBold = FRecursiveObject & {
  type: 'b'
}

type FItalic = FRecursiveObject & {
  type: 'i'
}

type FCode = {
  type: 'c'
  content: string
}

type FVerbatim = {
  type: 'v'
  content: string
}

type FStrikeThrough = FRecursiveObject & {
  type: '+'
}

type FUnderline = FRecursiveObject & {
  type: 'u'
}
type FSuperscript = FRecursiveObject & {
  type: '^'
}
type FSubscript = FRecursiveObject & {
  type: '_'
}
type FText = {
  type: 't'
  content: string
}
type FTimestamp = {
  type: 'Z'
  content: string
}
type FFootnoteReference = FRecursiveObject & {
  type: 'f'
  label: string
}
type FLatexFragment = {
  type: 'X'
  content: string
}
type FEntity = {
  type: '?'
  content: string
  html: string
}
type FTableCell = FRecursiveObject & {
  type: 'C'
}

type FObjectType =
  | FLink
  | FBold
  | FItalic
  | FCode
  | FVerbatim
  | FStrikeThrough
  | FUnderline
  | FSuperscript
  | FSubscript
  | FText
  | FTimestamp
  | FFootnoteReference
  | FLatexFragment
  | FEntity
  | FTableCell

type FHeading = FRecursiveObject & {
  type: 'h'
  level: number
  todoKeyword: string | null
  priority: string | null
  commented: boolean
  tags: string[]
}

type FParagraph = {
  type: 'p'
  content: FObjectType[]
}

type FElementType =
  | FHeading
  //  | FPlanning
  //  | FNodeProperty
  //  | FListItemTag
  //  | FCommentBlock
  //  | FSrcBlock
  //  | FExampleBlock
  //  | FExportBlock
  //  | FKeyword
  //  | FTableRow
  //  | FComment
  //  | FFixedWidth
  //  | FClock
  //  | FLatexEnvironment
  //  | FHorizontalRule
  //  | FDiarySexp
  | FParagraph

export type FDocument = {
  title?: string
  source?: string
  // TODO: Define Todo type? Has annotation (e.g.: comment, shortcut) removed?
  // TODO: Breakdown into list of Todo type items
  todoStates: Array<string>
  content: Array<FElementType>
}
