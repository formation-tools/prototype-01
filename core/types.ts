export type FRecursiveObject = {
  content: FObjectType[]
}

type FLink = FRecursiveObject & {
  type: 'a'
  target: string
  linkType: string
}

type FBold = FRecursiveObject & {
  type: 'b'
}

type FItalic = FRecursiveObject & {
  type: 'i'
}

type FCode = {
  type: 'c'
  content: string
}

type FVerbatim = {
  type: 'v'
  content: string
}

type FStrikeThrough = FRecursiveObject & {
  type: '+'
}

type FUnderline = FRecursiveObject & {
  type: 'u'
}
type FSuperscript = FRecursiveObject & {
  type: '^'
}
type FSubscript = FRecursiveObject & {
  type: '_'
}
type FText = {
  type: 't'
  content: string
}
// TODO: Extend timestamp with relevant fields
type FTimestamp = {
  type: 'Z'
  content: string
}
type FFootnoteReference = FRecursiveObject & {
  type: 'f'
  label: string
}
type FLatexFragment = {
  type: 'X'
  content: string
}
type FEntity = {
  type: '?'
  content: string
  html: string
}
type FTableCell = FRecursiveObject & {
  type: 'C'
}

export type FObjectType =
  | FLink
  | FBold
  | FItalic
  | FCode
  | FVerbatim
  | FStrikeThrough
  | FUnderline
  | FSuperscript
  | FSubscript
  | FText
  | FTimestamp
  | FFootnoteReference
  | FLatexFragment
  | FEntity
  | FTableCell

export type FHeading = FRecursiveObject & {
  type: 'h'
  level: number
  todoKeyword?: string
  priority?: string
  commented: boolean
  tags?: string[]
}

type FParagraph = FRecursiveObject & {
  type: 'p'
}

type FElementFallback = {
  type: 'e'
  content: string
}

export type FElement =
  | FHeading
  //  | FPlanning
  //  | FNodeProperty
  //  | FListItemTag
  //  | FCommentBlock
  //  | FSrcBlock
  //  | FExampleBlock
  //  | FExportBlock
  //  | FKeyword
  //  | FTableRow
  //  | FComment
  //  | FFixedWidth
  //  | FClock
  //  | FLatexEnvironment
  //  | FHorizontalRule
  //  | FDiarySexp
  | FParagraph
  | FElementFallback

type FGreaterElementFallback = {
  type: 'E'
  content: string
}

type FSection = {
  type: 'S'
  content: FElementType[]
}

export type FList = {
  type: 'L'
  variant: 'ordered' | 'unordered' | 'descriptive'
  content: FListItem[]
}

export type FListItem = {
  type: 'I'
  checkbox: 'on' | 'off' | 'trans' | null
  content: FElementType[]
}

// TODO: Expand
export type FGreaterElement =
  | FSection
  | FGreaterElementFallback
  | FList
  | FListItem

export type FElementType = FElement | FGreaterElement

export type FDocument = {
  title?: string
  source?: string
  // TODO: Define Todo type? Has annotation (e.g.: comment, shortcut) removed?
  // TODO: Breakdown into list of Todo type items
  todoStates: Array<string>

  // TODO: Compare Array vs Map
  // TODO: Design Map w/ a key-naming scheme for fast start/end/range lookup
  content: Array<FElementType>
}

export type FTableOfContentsEntry = {
  heading: FHeading
  // The text field should be rendered as the TOC text
  text: FObjectType[]
  plaintext: string
}
export type FTableOfContents = FTableOfContentsEntry[]

export type FNestedTableOfContentsEntry = {
  heading: FHeading
  // The text field should be rendered as the TOC text
  text: FObjectType[]
  plaintext: string
  children: FNestedTableOfContentsEntry[]
}
export type FNestedTableOfContents = FNestedTableOfContentsEntry[]

export const emptyDocument = {
  content: [],
  todoStates: [],
}
