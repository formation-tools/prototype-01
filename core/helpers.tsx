import base64url from 'base64url'

export const defaultTarget = encodeTarget(
  'https://gitlab.com/formation.tools/eng/engineering/-/raw/main/README.org',
)

export function encodeTarget(url: string) {
  return base64url.encode(url)
}
