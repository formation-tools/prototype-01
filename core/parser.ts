import { unified } from 'unified'
import parser from 'uniorg-parse'
// https://github.com/rasendubi/uniorg/blob/master/packages/uniorg/src/index.ts
// https://github.com/rasendubi/uniorg/blob/master/packages/uniorg-parse/src/parser.ts#L4
import { OrgData, ObjectType, GreaterElementType, ElementType } from 'uniorg'

import { FDocument, FObjectType, FElementType, emptyDocument } from './types'

// TODO: Potentially clean up by letting users import types directly
export type { FDocument }

function assertExhaustive(
  value: never,
  message: string = 'Reached unexpected case in exhaustive switch',
): never {
  throw new Error(message)
}

export function extractLabel(el: FObjectType): string {
  if ('content' in el) {
    switch (typeof el.content) {
      case 'string':
        return el.content
      default:
        return el.content
          .map(extractLabel)
          .reduce((next, acc) => acc.concat(next), '')
    }
  } else {
    assertExhaustive(el)
  }
}

function mapObjectType(x: ObjectType): FObjectType {
  // TODO: Restructure to a list of non-string members
  // A lot of information is lost in just reducing this to strings
  switch (x.type) {
    case 'link':
      // https://orgmode.org/worg/dev/org-syntax.html#Links
      return {
        type: 'a',
        target: x.rawLink,
        linkType: x.linkType,
        content: x.children.map(mapObjectType),
      }
    case 'bold':
      return { type: 'b', content: x.children.map(mapObjectType) }
    case 'italic':
      return { type: 'i', content: x.children.map(mapObjectType) }
    case 'strike-through':
      return { type: '+', content: x.children.map(mapObjectType) }
    case 'underline':
      return { type: 'u', content: x.children.map(mapObjectType) }
    case 'superscript':
      return { type: '^', content: x.children.map(mapObjectType) }
    case 'subscript':
      return { type: '_', content: x.children.map(mapObjectType) }
    case 'code':
      return { type: 'c', content: x.value }
    case 'verbatim':
      return { type: 'v', content: x.value }
    case 'text':
      return { type: 't', content: x.value }
    case 'timestamp':
      return { type: 'Z', content: x.rawValue }
    case 'footnote-reference':
      return {
        type: 'f',
        label: x.label,
        content: x.children.map(mapObjectType),
      }
    case 'latex-fragment':
      return { type: 'X', content: x.value }
    case 'entity':
      return { type: '?', content: x.name, html: x.html }
    case 'table-cell':
      return { type: 'C', content: x.children.map(mapObjectType) }
    default:
      return assertExhaustive(x)
  }
}

function unpackElementType(
  x: GreaterElementType | ElementType,
): FElementType[] {
  switch (x.type) {
    // GreaterElementType
    case 'org-data':
      return []
    case 'section':
      return x.children.flatMap(unpackElementType)
    case 'property-drawer':
    case 'drawer':
    case 'plain-list':
    case 'list-item':
    case 'quote-block':
    case 'verse-block':
    case 'center-block':
    case 'special-block':
    case 'footnote-definition':
    case 'table':
      return []
    // ElementType
    case 'headline':
      return [
        {
          type: 'h',
          level: x.level,
          todoKeyword: x.todoKeyword,
          commented: x.commented,
          priority: x.priority,
          tags: x.tags,
          content: x.children.map(mapObjectType),
        },
      ]
    case 'planning':
    case 'node-property':
    case 'list-item-tag':
    case 'comment-block':
    case 'src-block':
    case 'example-block':
    case 'export-block':
    case 'keyword':
    case 'table-row':
    case 'comment':
    case 'fixed-width':
    case 'clock':
    case 'latex-environment':
    case 'horizontal-rule':
    case 'diary-sexp':
      return []
    case 'paragraph':
      return [{ type: 'p', content: x.children.map(mapObjectType) }]
    default:
      return assertExhaustive(x)
  }
}

// Convert document to internal representation
function convert(
  acc: FDocument,
  node: GreaterElementType | ElementType,
  idx: number,
): FDocument {
  switch (node.type) {
    // GreaterElementType
    case 'org-data':
      return node.children.reduce(convert, acc)
    case 'section':
      return {
        ...acc,
        content: [...acc.content, ...node.children.flatMap(unpackElementType)],
      }
    case 'property-drawer':
    case 'drawer':
    case 'plain-list':
    case 'list-item':
    case 'quote-block':
    case 'verse-block':
    case 'center-block':
    case 'special-block':
    case 'footnote-definition':
    case 'table':
      return acc // noop
    // ElementType
    case 'headline':
      return {
        ...acc,
        content: [...acc.content, ...unpackElementType(node)],
      }
    case 'planning':
    case 'node-property':
    case 'list-item-tag':
    case 'comment-block':
    case 'src-block':
    case 'example-block':
    case 'export-block':
      return acc // noop
    case 'keyword':
      switch (node.key) {
        case 'TITLE':
          return { ...acc, title: node.value }
        case 'TODO':
          // FIXME: Accomodate for multiple swimlanes as per https://orgmode.org/manual/Per_002dfile-keywords.html
          return {
            ...acc,
            todoStates: node.value.split(' ').filter((x) => x != '|'),
          }
        default:
          return acc
      }
    case 'table-row':
    case 'comment':
    case 'fixed-width':
    case 'clock':
    case 'latex-environment':
    case 'horizontal-rule':
    case 'diary-sexp':
      return acc // noop
    case 'paragraph':
      return {
        ...acc,
        content: [...acc.content, ...unpackElementType(node)],
      }
    default:
      return assertExhaustive(node)
  }
}

export default function parse(text: string): FDocument {
  const ast = unified().use(parser).parse(text) as OrgData
  return convert(emptyDocument, ast, 0)
}
