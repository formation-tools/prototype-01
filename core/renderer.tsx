import { Fragment } from 'react'
import {
  FDocument,
  FHeading,
  FLink,
  FElementType,
  FObjectType,
} from 'core/types'

import Code, { CodeProps } from 'components/doc/Code'
import Heading, { HeadingProps } from 'components/doc/Heading'
import Paragraph, { ParagraphProps } from 'components/doc/Paragraph'
import Link, { LinkProps } from 'components/doc/Link'
import FallbackInline, {
  FallbackInlineProps,
} from 'components/doc/FallbackInline'
import FallbackBlock, { FallbackBlockProps } from 'components/doc/FallbackBlock'
import Date, { DateProps } from 'components/doc/Date'
import { List, ListChild, ListChildProps, ListProps } from 'components/doc/List'

function assertExhaustive(
  value: never,
  message: string = 'Reached unexpected case in exhaustive switch',
): never {
  throw new Error(message)
}

export function renderElement(
  el: FElementType,
  i: number | string,
  // FIXME: The optional doc arg is a nasty hack to get context into the
  // rendering stage
  doc?: FDocument,
): JSX.Element[] {
  switch (el.type) {
    case 'S':
      // TODO: Render section in collapsible component
      return el.content.flatMap((el, idx) =>
        renderElement(el, `S${i}-${idx}`, doc),
      )
    case 'L':
      return [
        <List key={`L${i}`}>
          {el.content.flatMap((x, idx) =>
            renderElement(x, `L${i}-I${idx}`, doc),
          )}
        </List>,
      ]

    case 'I':
      if (el.content.length === 0) {
        return []
      }

      const [head, ...rest] = el.content

      // We don't know how to render a list if the head is not a paragraph
      if (head.type !== 'p') {
        return []
      }

      const headObjects = head.content.flatMap((x, idx) =>
        renderObject(x, `${i}-label-${idx}`, doc),
      )

      const restObjects =
        rest && rest.length > 0
          ? rest.flatMap((x, idx) => renderElement(x, `${i}-body=${idx}`, doc))
          : []

      return [
        <ListChild key={i} label={headObjects}>
          {restObjects}
        </ListChild>,
      ]

    case 'h':
      // TODO: @vidbina Source document context such that we can retrieve id slug

      switch (el.level) {
        case 1:
        case 2:
        case 3:
        case 4:
        case 5:
        case 6:
          const { id, level, todoKeyword, priority, commented, tags } = el
          const renderedId = doc ? doc.headingIdToSlugIndex[id] || id : id

          return [
            // TODO: Migrate TODOs here to Heading component source file
            // TODO: Implement heading keyword
            // TODO: Implement heading priority
            // TODO: Implement heading comment status
            // TODO: Implement heading tags

            <Heading
              id={renderedId}
              key={`h${i}`}
              level={level}
              todoKeyword={todoKeyword}
              priority={priority}
              commented={commented}
              tags={tags}
            >
              {el.content.flatMap((el, idx) =>
                renderObject(el, `h${i}-${idx}`, doc),
              )}
            </Heading>,
          ]
        default:
          // FIXME: Implement a fallback case
          return []
      }
    case 'p':
      return [
        <Paragraph key={`p${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `p${i}-${idx}`, doc),
          )}
        </Paragraph>,
      ]
    case 'E':
    case '#':
    case '/':
      return [<FallbackBlock key={`E${i}`}>{el.content}</FallbackBlock>]
    case 'e':
      return [<FallbackInline key={`e${i}`}>{el.content}</FallbackInline>]
    default:
      assertExhaustive(el)
  }
}

export function destinationForLink(link: FLink, doc?: FDocument): string {
  if (doc === undefined) {
    return link.target
  }

  switch (link.linkType) {
    case 'fuzzy':
      if (link.target.charAt(0) !== '*') {
        return link.target
      }
      const fuzzyTarget = link.target.slice(1)

      const id = doc.headingFuzzyToIdIndex[fuzzyTarget]
      if (id === undefined) {
        return link.target
      }

      const headingSlug = doc.headingIdToSlugIndex[id]
      if (headingSlug === undefined) {
        return link.target
      }

      return `/#${headingSlug}`
    case 'id':
      // TODO: Implement
      return link.target
    default:
      return link.target
  }
}

export function renderObject(
  el: FObjectType,
  i: number | string,
  // FIXME: Reevaluate if this context-dump is really necessary
  doc?: FDocument,
): JSX.Element[] {
  switch (el.type) {
    case 'a':
      return [
        <Link
          key={`a${i}`}
          url={destinationForLink(el, doc)}
          linkType={el.linkType}
          label={el.content.flatMap((el, idx) =>
            renderObject(el, `a${i}-${idx}`, doc),
          )}
        />,
      ]
    case 'b':
      return [
        <b key={`b${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `b${i}-${idx}`, doc),
          )}
        </b>,
      ]
    case 'i':
      return [
        <i key={`i${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `i${i}-${idx}`, doc),
          )}
        </i>,
      ]
    case 'c':
      return [<code key={`c${i}`}>{el.content}</code>]
    case 'v':
      return [<code key={`v${i}`}>{el.content}</code>]
    case '+':
      return [
        <s key={`+${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `+${i}-${idx}`, doc),
          )}
        </s>,
      ]
    case 'u':
      return [
        <u key={`u${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `u${i}-${idx}`, doc),
          )}
        </u>,
      ]
    case '^':
      return [
        <sup key={`^${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `^${i}-${idx}`, doc),
          )}
        </sup>,
      ]
    case '_':
      return [
        <sub key={`_${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `_${i}-${idx}`, doc),
          )}
        </sub>,
      ]
    case 't':
      return [<Fragment key={`t${i}`}>{el.content}</Fragment>]
    case 'Z':
      // TODO: Pass along datetime
      return [
        <time key={`Z${i}`} dateTime="2015-10-21">
          {el.content}
        </time>,
      ]
    case 'f':
      // TODO: Implement footnote
      return [
        <em key={`f${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `f${i}-${idx}`, doc),
          )}
        </em>,
      ]
    case 'X':
      // TODO: Display LaTeX (using MathJaX or something server-side-rendered for better perf)
      return [<code key={`X${i}`}>{el.content}</code>]
    case '?':
      // TODO: Decide on and implement fallback-strategy for entity types
      return [<code key={`?${i}`}>{el.content}</code>]
    case 'C':
      return [
        <td key={`C${i}`}>
          {el.content.flatMap((el, idx) =>
            renderObject(el, `C${i}-${idx}`, doc),
          )}
        </td>,
      ]
    default:
      assertExhaustive(el)
  }
}
