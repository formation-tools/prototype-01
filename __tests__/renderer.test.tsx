import { describe, expect, it } from 'vitest'

import { render as r } from '@testing-library/react'

import { render, renderElement } from 'core/renderer'

import { emptyDocument } from 'core/types'

const f = (a) => {
  const { asFragment, getByText } = r(a)
  return asFragment()
}

describe('heading', () => {
  it('renders', () => {
    expect(
      f(
        renderElement({
          type: 'h',
          level: 1,
          todoKeyword: 'TODO',
          priority: 'A',
          commented: true,
          tags: ['idea', 'strategy'],
          content: [{ type: 't', content: 'Collect underpants' }],
        }),
      ),
    ).toMatchInlineSnapshot(`
      <DocumentFragment>
        <h1
          class="StyledHeading-sc-1rdh4aw-0 gLPrrO"
        >
          Collect underpants
        </h1>
      </DocumentFragment>
    `)
  })
})

describe('paragraph', () => {
  it('renders', () => {
    expect(
      f(
        renderElement({
          type: 'p',
          content: [
            { type: 't', content: 'The first part of the ' },
            { type: 'b', content: [{ type: 't', content: 'masterplan' }] },
          ],
        }),
      ),
    ).toMatchInlineSnapshot(`
      <DocumentFragment>
        <p
          class="StyledParagraph-sc-tbetod-0 lacipA"
        >
          The first part of the 
          <b>
            masterplan
          </b>
        </p>
      </DocumentFragment>
    `)
  })

  describe('including timestamp', () => {
    it('renders', () => {
      expect(
        f(
          renderElement({
            content: [
              {
                content: 'Watching Penelope and Morgan solve crimes on ',
                type: 't',
              },
              {
                content: '<2022-08-14 Sun>',
                type: 'Z',
              },
              {
                content: ' while testing fallback components.',
                type: 't',
              },
            ],
            type: 'p',
          }),
        ),
      ).toMatchInlineSnapshot(`
        <DocumentFragment>
          <p
            class="StyledParagraph-sc-tbetod-0 lacipA"
          >
            Watching Penelope and Morgan solve crimes on 
            <time
              datetime="2015-10-21"
            >
              &lt;2022-08-14 Sun&gt;
            </time>
             while testing fallback components.
          </p>
        </DocumentFragment>
      `)
    })
  })
})

describe('fallback', () => {
  describe('for element', () => {
    it('renders', () => {
      expect(
        f(
          renderElement({
            type: 'e',
            content: 'some random junk',
          }),
        ),
      ).toMatchInlineSnapshot(`
        <DocumentFragment>
          <code
            class="p-2 rounded bg-gray-300/30 "
          >
            some random junk
          </code>
        </DocumentFragment>
      `)
    })
  })
  describe('for greater element', () => {
    it('renders', () => {
      expect(
        f(
          renderElement({
            type: 'E',
            content: 'some random junk',
          }),
        ),
      ).toMatchInlineSnapshot(`
        <DocumentFragment>
          <pre
            class="overflow-auto p-2 rounded bg-gray-300/30 "
          >
            some random junk
          </pre>
        </DocumentFragment>
      `)
    })
  })
})
