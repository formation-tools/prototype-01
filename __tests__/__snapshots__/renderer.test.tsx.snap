// Vitest Snapshot v1

exports[`heading > renders with id when slug is not present 1`] = `
<DocumentFragment>
  <div
    class="p-2 md:p-4 transition duration-0 hover:duration-150 hover:bg-gray-100 target:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 target:ring-1 hover:ring-gray-300 target:ring-gray-300 md:flex relative text-2xl group false"
    id="hopefully-random-id"
  >
    <a
      class="align-middle inline-block inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis font-bold hover:text-c-blue-hover invisible group-hover:visible"
      href="/#hopefully-random-id"
    >
      <svg
        aria-hidden="true"
        class="h-5 rotate-45 inline"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.232 4.232a2.5 2.5 0 013.536 3.536l-1.225 1.224a.75.75 0 001.061 1.06l1.224-1.224a4 4 0 00-5.656-5.656l-3 3a4 4 0 00.225 5.865.75.75 0 00.977-1.138 2.5 2.5 0 01-.142-3.667l3-3z"
        />
        <path
          d="M11.603 7.963a.75.75 0 00-.977 1.138 2.5 2.5 0 01.142 3.667l-3 3a2.5 2.5 0 01-3.536-3.536l1.225-1.224a.75.75 0 00-1.061-1.06l-1.224 1.224a4 4 0 105.656 5.656l3-3a4 4 0 00-.225-5.865z"
        />
      </svg>
    </a>
    <span
      class="inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis flex-none"
    >
      <span
        class="border border-1 rounded bg-red-100 text-red-900 border-red-300 p-0 px-2 "
      >
        TODO
      </span>
    </span>
    <span
      class="inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis font-bold grow"
    >
      <h1
        class="inline"
        id="hopefully-random-id"
      >
        Collect underpants
      </h1>
    </span>
    <span
      class="inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis space-x-1 space-x-reverse space-x-y space-y-reverse flex flex-row-reverse flex-wrap w-2/6"
    >
      <span
        class="border border-1 rounded-full bg-blue-100 text-blue-900 border-blue-300 text-xs h-6 py-1 px-2 "
      >
        idea
      </span>
      <span
        class="border border-1 rounded-full bg-blue-100 text-blue-900 border-blue-300 text-xs h-6 py-1 px-2 "
      >
        strategy
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`heading > renders with slug as id when present 1`] = `
<DocumentFragment>
  <div
    class="p-2 md:p-4 transition duration-0 hover:duration-150 hover:bg-gray-100 target:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 target:ring-1 hover:ring-gray-300 target:ring-gray-300 md:flex relative text-2xl group false"
    id="some-slug"
  >
    <a
      class="align-middle inline-block inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis font-bold hover:text-c-blue-hover invisible group-hover:visible"
      href="/#some-slug"
    >
      <svg
        aria-hidden="true"
        class="h-5 rotate-45 inline"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.232 4.232a2.5 2.5 0 013.536 3.536l-1.225 1.224a.75.75 0 001.061 1.06l1.224-1.224a4 4 0 00-5.656-5.656l-3 3a4 4 0 00.225 5.865.75.75 0 00.977-1.138 2.5 2.5 0 01-.142-3.667l3-3z"
        />
        <path
          d="M11.603 7.963a.75.75 0 00-.977 1.138 2.5 2.5 0 01.142 3.667l-3 3a2.5 2.5 0 01-3.536-3.536l1.225-1.224a.75.75 0 00-1.061-1.06l-1.224 1.224a4 4 0 105.656 5.656l3-3a4 4 0 00-.225-5.865z"
        />
      </svg>
    </a>
    <span
      class="inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis flex-none"
    >
      <span
        class="border border-1 rounded bg-red-100 text-red-900 border-red-300 p-0 px-2 "
      >
        TODO
      </span>
    </span>
    <span
      class="inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis font-bold grow"
    >
      <h1
        class="inline"
        id="some-slug"
      >
        Collect underpants
      </h1>
    </span>
    <span
      class="inline-block align-baseline align-text-bottom mr-3 last:mr-0 text-ellipsis space-x-1 space-x-reverse space-x-y space-y-reverse flex flex-row-reverse flex-wrap w-2/6"
    >
      <span
        class="border border-1 rounded-full bg-blue-100 text-blue-900 border-blue-300 text-xs h-6 py-1 px-2 "
      >
        idea
      </span>
      <span
        class="border border-1 rounded-full bg-blue-100 text-blue-900 border-blue-300 text-xs h-6 py-1 px-2 "
      >
        strategy
      </span>
    </span>
  </div>
</DocumentFragment>
`;
